src/lib/list.c

```diff
@@ -305,9 +305,11 @@ bool list_longer_than(struct list *list, int count)
 
 void *list_search(struct list *list, const void *item, comparison_fn_t comparison_fn)
 {
-	for (; list; list = list->next) {
-		if (comparison_fn(list->data, item) == 0) {
-			return list->data;
+	struct list *heystack = list_head(list);
```

Comment:
Usually list implementations consider the current pointer the begging of the list. The list implementation on swupd has an unusual (and inefficient) behavior of always assuming that the user can point to any element of the list. Because of that in several functions we don't care where we are pointing, what I consider a bad habit. Instead of going to the begging of the list on every single operation I prefer to try to remove the list_head() calls as much as possible and try to use the list in a saner way.

By saying that this is not a bug in the search function. Maybe the documentation should be updated to say that the search function starts to look for the item in the pointed element. And caller is in charge on pointing to correct item.

Reply:
but what is the harm with making sure you are at the beginning of the list by doing a list_head? if you are already at the beginning it won't do anything so it won't have performance impacts, I agree we should follow some good practices when using lists, but if for some reason we are not I'd like functions to be robust to catch it. Otherwise we could end up having bugs that would be very difficult to debug

Reply:
The major harm is that we need to add list_head() calls everywhere. In any point of your code that you receive a list as an argument you need to call to list_head() because you never know if the list is in the begging or at the end. The problem that you are trying to fix is that the called of list_search is pointing to an incorrect position of this list and not that list_search is not working.
If instead of list search you were calling other function that works with a list you would have the same problem.

Reply:
yes, I understand that, but still don't see any harm on making sure. But I'll remove this commit if you think is not necessary

Reply:
Ok, for now let's leave it here, but let's try to make sure we are always returning the head of the list on functions too. To remove all unnecessary calls to list_head() sometime in the future

Reply:
ok